internal/store/ring.go

```diff
@@ -0,0 +1,226 @@
+)
+
+var ringSelect sq.SelectBuilder
+
+func init() {
+	ringSelect = sq.
+		Select("Ring.ID", "Name", "Priority", "InstallationGroup", "SoakTime", "Image", "Version", "Provisioner", "State", "CreateAt", "DeleteAt", "ReleaseAt", "APISecurityLock", "LockAcquiredBy", "LockAcquiredAt").
+		From("Ring")
+}
+
+type rawRing struct {
+	*model.Ring
+}
+
+type rawRings []*rawRing
+
+func (r *rawRing) toRing() (*model.Ring, error) {
+	return r.Ring, nil
+}
+
+func (rc *rawRings) toRings() ([]*model.Ring, error) {
+	var rings []*model.Ring
+	for _, rawRing := range *rc {
+		ring, err := rawRing.toRing()
+		if err != nil {
+			return nil, err
+		}
+		rings = append(rings, ring)
+	}
+
+	return rings, nil
+}
+
+// GetRing fetches the given ring by id.
+func (sqlStore *SQLStore) GetRing(id string) (*model.Ring, error) {
+	var rawRing rawRing
+	err := sqlStore.getBuilder(sqlStore.db, &rawRing, ringSelect.Where("ID = ?", id))
+	if err == sql.ErrNoRows {
+		return nil, nil
+	} else if err != nil {
+		return nil, errors.Wrap(err, "failed to get ring by id")
+	}
+
+	return rawRing.toRing()
+}
+
+// GetRings fetches the given page of created rings. The first page is 0.
+func (sqlStore *SQLStore) GetRings(filter *model.RingFilter) ([]*model.Ring, error) {
+	builder := ringSelect.
+		OrderBy("CreateAt ASC")
+	builder = sqlStore.applyRingsFilter(builder, filter)
+
+	var rawRings rawRings
+	err := sqlStore.selectBuilder(sqlStore.db, &rawRings, builder)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to query for rings")
+	}
+
+	return rawRings.toRings()
+}
+
+func (sqlStore *SQLStore) applyRingsFilter(builder sq.SelectBuilder, filter *model.RingFilter) sq.SelectBuilder {
+	if filter.PerPage != model.AllPerPage {
+		builder = builder.
+			Limit(uint64(filter.PerPage)).
+			Offset(uint64(filter.Page * filter.PerPage))
+	}
+
+	if !filter.IncludeDeleted {
+		builder = builder.Where("DeleteAt = 0")
+	}
+
+	return builder
+}
+
+// GetUnlockedRingsPendingWork returns an unlocked ring in a pending state.
+func (sqlStore *SQLStore) GetUnlockedRingsPendingWork() ([]*model.Ring, error) {
+	builder := ringSelect.
+		Where(sq.Eq{
+			"State": model.AllRingStatesPendingWork,
+		}).
+		Where("LockAcquiredAt = 0").
+		OrderBy("CreateAt ASC")
+
+	var rawRings rawRings
+	err := sqlStore.selectBuilder(sqlStore.db, &rawRings, builder)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to query for rings")
+	}
+
+	return rawRings.toRings()
+}
+
+// CreateRing records the given ring to the database, assigning it a unique ID.
+func (sqlStore *SQLStore) CreateRing(ring *model.Ring) error {
+	tx, err := sqlStore.beginTransaction(sqlStore.db)
+	if err != nil {
+		return errors.Wrap(err, "failed to begin transaction")
+	}
+	defer tx.RollbackUnlessCommitted()
+
+	if err = sqlStore.createRing(tx, ring); err != nil {
+		return errors.Wrap(err, "failed to create ring")
+	}
+
+	if err = tx.Commit(); err != nil {
+		return errors.Wrap(err, "failed to commit the transaction")
+	}
```

Comment:
When running only one DB operation you do not need to start transaction separately, so you could just do `sqlStore.createRing(sqlStore.db, ring)` unless you want to add more things there in the future.